[IMP] base: ir.cron documentation (#12077)
odoo/documentation: content/developer/reference/backend/actions.rst

Writing CRON functions
----------------------

When running a scheduled action, it's recommended that you try to batch the
progress in order to a worker for a long period of time and possibly run into
timeout exceptions. Therefore you should split the processing so that each call
make progress on some of the work to be done.

When writing a such a function, you should focus on processing a single batch.
A batch should process one or many records and should generally take no more
than *a few seconds*.

.. code-block:: python

    def _cron_do_something(self, *, limit=300):  # limit: allows for tweaking
        domain = [('state', '=', 'ready')]
        records = self.search(domain, limit=limit)
        records.do_something()
        # notify progression
        remaining = 0 if len(records) == limit else self.search_count(domain)
        self.env['ir.cron']._commit_progress(len(records), remaining=remaining)

Work is committed by the framework after each batch. The framework will
call the function as many times as necessary to process the remaining work.

If called from outside of the cron job, the progress function call will have
no effect.

> There are some cases, where you may want to keep resources between multiple
> batches or want to loop yourself to handle exceptions.
> If this is the case, you must notify the scheduler of the progress of your work
> by calling ``ir.cron._commit_progress`` and checking the result. The progress
> function returns the number of seconds remaining for the call; if it is 0, you
> must return as soon as possible.

Reviewer: ```suggestion
In some cases, you may want to save resources between multiple
batches or manage the loop yourself to handle exceptions.
In this case, you should inform the scheduler of the progress of your work
by calling `ir.cron._commit_progress` and checking the result. The progress
function returns the number of seconds remaining for the call; if it is 0, you
must return as soon as possible.
```

> must return as soon as possible.

Reviewer: What do you mean by "as soon as possible"? Not to be nitpicky, but it could be interpreted in widely different ways. "immediately"? "without any new DB access"?

Author: It is intentionally left open to interpretation. At some point, you will reach the hard limit and be killed.

Example: Some jobs may do a logical batch consisting of : loop(A, commit, B, commit). When after A they have no time left, they may decide to still do B to have consistent results, but they should not continue looping. If they want, they may continue  running at he risk of being killed and unfair to other jobs.

> The following is an example of how to commit after each record that is

Author: Ok. I didn't intended to break the paragraph, it's was just easier to read/change without moving the whole paragraph contents.

        domain = [('state', '=', 'ready')]
        records = self.search(domain)
        self.env['ir.cron']._commit_progress(remaining=len(records))

        with open_some_connection() as conn:
            for record in records:
                # You may have other needs; we do some common stuff here:
                # - lock record (also checks existence)
                # - prefetch: break prefetch in this case, we process one record
                # - filtered_domain: record may have changed
                record = record.try_lock_for_update().filtered_domain(domain)
                if not record:
                    continue
                # Processing the batch here...
                try
                    record.do_something(conn)
                    if not self.env['ir.cron']._commit_progress(1):
                        break
                except Exception:
                    # if you handle exceptions, the default stategy is to
                    # rollback first the error
                    self.env.cr.rollback()
                    _logger.warning(...)
                    # you may commit some status using _commit_progress

> Running CRON functions

Reviewer: Why uppercase cron?

----------------------

You should not call CRON functions directly.
There are two ways to run functions: immediately (in the current thread, but
still in a separate transaction) or trigger to start at a specific time or as
soon as possible.

.. code-block:: python

    # run now (still in a separate transaction)
    cron_record.method_direct_trigger()
    # trigger

>     when: datetime | None = None  # for now

Reviewer: ```suggestion
    when: datetime | None = None  # None = as soon as possible
```

    cron_record._trigger(when)

> Testing of a CRON function should be done by calling ``method_direct_trigger``

Reviewer: Document `method_direct_trigger`, I thought it was an example name.
